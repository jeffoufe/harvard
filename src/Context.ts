import React, { SetStateAction, Dispatch } from "react";

interface Context {
    isMenuOpen: boolean,
    setMenuOpen: Dispatch<SetStateAction<boolean>>
}

export default React.createContext<Context>({} as Context)