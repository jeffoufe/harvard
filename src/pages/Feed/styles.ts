import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

export default makeStyles((theme: Theme) =>
    createStyles({
        initialLoading: {
            position: 'absolute',
            left: 'calc(50vw - 25px)',
            top: 'calc(50vh)',
            transition: theme.transitions.create(['left'], {
                easing: theme.transitions.easing.sharp,
                duration: theme.transitions.duration.leavingScreen,
            }),
        },
        initialLoadingShift: {
            left: 'calc(50vw - 25px + 120px)',
        },
        grid: {
            flexGrow: 1
        },
        paper: {
            padding: theme.spacing(2),
            textAlign: 'center',
            color: theme.palette.text.secondary,
            marginBottom: 10
        },
        printImage: {
            maxHeight: 200,
            maxWidth: '100%'
        }
    })
)