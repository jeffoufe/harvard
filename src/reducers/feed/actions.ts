import { Print, FeedActionSuccess, FeedActionFailure, FeedActionLoading, } from './types';
import {
    FEED_LOADING,
    FEED_SUCCESS,
    FEED_FAILURE,
    FEED_FETCH
} from './constants';

export const fetchFeed = () => ({
    type: FEED_FETCH
})

export const feedLoading = (): FeedActionLoading => ({
    type: FEED_LOADING
});

export const feedSuccess = (prints: Array<Print>, nextUrl: string | null): FeedActionSuccess => ({
    type: FEED_SUCCESS,
    payload: {
        prints, 
        nextUrl
    }
});

export const feedFailure = (error: string): FeedActionFailure => ({
    type: FEED_FAILURE,
    payload: {
        error
    }
});