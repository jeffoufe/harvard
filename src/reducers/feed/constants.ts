export const INITIAL_URL = 'https://api.harvardartmuseums.org/objec?classification=Prints&hasimage=1&apikey=c28e4be0-4c0e-11ea-90d6-25d9a9fe80fc&size=10&page=1'

export const INITIAL_STATE = {
    loading: true,
    prints: [],
    nextUrl: null,
    error: null
}

export const FEED_FETCH = 'FEED/FEED_FETCH';
export const FEED_LOADING = 'FEED/FEED_LOADING';
export const FEED_SUCCESS = 'FEED/FEED_SUCCESS';
export const FEED_FAILURE = 'FEED/FEED_FAILURE';