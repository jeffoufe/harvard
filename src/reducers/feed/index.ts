
import { FeedState, FeedAction } from './types';
import {
    INITIAL_STATE,
    FEED_LOADING,
    FEED_SUCCESS,
    FEED_FAILURE
} from './constants';

export default (state: FeedState = INITIAL_STATE, action: FeedAction) => {
    switch (action.type) {
        case FEED_LOADING:
            return {
                ...state,
                loading: true,
                error: null
            };
        case FEED_SUCCESS:
            return {
                ...state,
                loading: false,
                prints: [...state.prints, ...action.payload.prints],
                nextUrl: action.payload.nextUrl
            };
        case FEED_FAILURE:
            return {
                ...state,
                loading: false,
                error: action.payload.error
            }
        default:
            return state;
    }
};