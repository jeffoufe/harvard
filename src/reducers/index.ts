export { default as feedReducer } from './feed';