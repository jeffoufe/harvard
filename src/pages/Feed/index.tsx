import React, { useEffect, useContext } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { MissingImage } from '../../components';
import { fetchFeed } from '../../reducers/feed/actions';
import { FeedState, Print } from '../../reducers/feed/types';
import { CircularProgress, Grid, Paper } from '@material-ui/core';
import useStyles from './styles';
import Context from '../../Context';
import clsx from 'clsx';

const Feed = () => {
    const dispatch = useDispatch();
    const classes = useStyles();
    const [isMenuOpen] = useContext(Context);
    const { prints, loading } = useSelector((state: FeedState) => state);

    useEffect(() => {
        dispatch(fetchFeed())
    }, [dispatch]);

    return <>
        {loading && !prints && (
            <div className={clsx(classes.initialLoading, isMenuOpen && classes.initialLoadingShift)}>
                <CircularProgress />
            </div>
        )}
        {!!prints.length && (
            <div className={classes.grid}>
                <Grid container spacing={3}>
                    {prints.map((print: Print) => (
                        <Grid item xs={12} sm={6} md={4} lg={3} direction={"row"}>
                            <Paper className={classes.paper}>
                                {!print.images.length 
                                    ? <MissingImage />
                                    : (
                                        <img 
                                            alt={print.title}
                                            src={(print.images[0] || {}).baseimageurl}
                                            className={classes.printImage}
                                        />
                                    )
                                }
                                <div>{print.title}</div>
                            </Paper>
                        </Grid>
                    ))}
                </Grid>
            </div>
        )}
    </>
}

Feed.displayName = 'Feed';
export default Feed;