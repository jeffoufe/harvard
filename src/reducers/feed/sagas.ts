import { takeEvery, put, select, call } from 'redux-saga/effects';
import { INITIAL_URL, FEED_FETCH } from './constants';
import { FeedState } from './types';
import { fetchItems } from './utils';
import {
    feedLoading,
    feedSuccess,
    feedFailure
} from './actions';

export const nextUrlSelector  = (state: FeedState) => state.nextUrl;

export function* loadItems() {
    try {
        yield put(feedLoading())
        const nextUrl = yield select(nextUrlSelector);
        const responseJson = yield call(fetchItems, nextUrl || INITIAL_URL);
        yield put(feedSuccess(responseJson.records, responseJson.info.next))
    } catch (error) {
        yield put(feedFailure('Failed to fetch'));
    }
} 

export function* watchLoadItems() {
    // @ts-ignore
    yield takeEvery(FEED_FETCH, loadItems);
}