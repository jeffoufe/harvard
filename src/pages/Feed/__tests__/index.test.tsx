import React from 'react';
import { createMount } from '@material-ui/core/test-utils';
import Feed from '..'
import * as redux from "react-redux";
import { createStore } from 'redux';
import { fetchFeed } from '../../../reducers/feed/actions';
import { feedReducer } from '../../../reducers';
import { INITIAL_STATE } from '../../../reducers/feed/constants';
import Context from '../../../Context';

describe('Feed', () => {
    const mockDispatchFn = jest.fn()

    const store = createStore(feedReducer, INITIAL_STATE);

    jest.spyOn(React, 'useEffect').mockImplementation((f: any) => f());
    jest.spyOn(redux, 'useDispatch').mockReturnValue(mockDispatchFn);
    jest.spyOn(redux, 'useSelector').mockReturnValue({
        prints: store.getState().prints,
        loading: store.getState().loading
    });

    let mount;
    let wrapper;

    beforeAll(() => {
        mount = createMount();
    });
    
    afterAll(() => {
        mount.cleanUp();
    });

    it('Renders correctly', () => {
        wrapper = mount(
            <Context.Provider value={[true]} >
                <Feed />
            </Context.Provider>
        );
        expect(wrapper).toMatchSnapshot();
    })

    it('UseEffect', () => {
        expect(mockDispatchFn).toHaveBeenCalledWith(fetchFeed())
    })
})