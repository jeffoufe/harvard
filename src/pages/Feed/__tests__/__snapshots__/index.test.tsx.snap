// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed Renders correctly 1`] = `<Feed />`;
